File: src-js/VariationSet.js
var VariationSet = (function(){
	
	var VariationSet = function () {
		CombinatoricsBaseSet.apply( this, arguments );
	}
	VariationSet.prototype = new CombinatoricsBaseSet();
	VariationSet.prototype.constructor = VariationSet;
	
	VariationSet.prototype.rewind = function () {
		
		this.current = BigInteger( BigInteger.ZERO );
		this.totalResults = BigInteger( BigInteger.ZERO );
		var v = new Array( this.to-this.from+1 );
		
		for ( var i = 0, k = this.to-this.from; i <= k; i++ )
		{
			v[i] = new Variation( this.elements, i );
			this.totalResults = this.totalResults.add( v[i].total() );
		}
		
		this.generator = new Variation( this.elements, this.from );
		this.current = 0;
		this.currentSet = this.from;
	}
	
	/* CombinatoricsBase */
	VariationSet.prototype.nextGenerator = function () {
		
		this.generator = new Variation( this.elements, this.currentSet );
		return this.generator;
	}
	
	return VariationSet;
	
})();

File: src/de/bezier/math/combinatorics/VariationSet.java
package de.bezier.math.combinatorics;

import java.math.BigInteger;

/**
 *	A range of Variations
 *
 *	<p>Represents a range of Variations. See intro to CombinationSet for an explaination of what that means.</p>
 *
 *	@see de.bezier.math.combinatorics.CombinationSet
 */
public class VariationSet
extends CombinatoricsBaseSet
{
	public VariationSet ( int elements ) { super(elements); }
	public VariationSet ( int elements, int from, int to ) { super(elements, from, to); }
	
	/**
	 *	Does what it says: rewinds the internal state to start over.
	 */
	public void rewind ()
	{
		current = BigInteger.ZERO;
		totalResults = BigInteger.ZERO;
		Variation v[] = new Variation[to-from+1];
		for ( int i = from; i <= to; i++ )
		{
			v[i] = new Variation(elements, i);
			totalResults = 
				totalResults.add( v[i].total() );
		}
		
		generator = new Variation(elements, from);
		current = BigInteger.ZERO;
		currentSet = from;
	}
	
	CombinatoricsBase nextGenerator ()
	{
		generator = new Variation(elements, currentSet);
		return generator;
	}
}
